File: Algorithms/Scala/0094/Solution.scala
/**
 * Definition for a binary tree node.
 * class TreeNode(var _value: Int) {
 *   var value: Int = _value
 *   var left: TreeNode = null
 *   var right: TreeNode = null
 * }
 */
object Solution {
    def inorderTraversal(root: TreeNode): List[Int] = {
        if (root == null) List.empty[Int]
        else inorderTraversal(root.left) ::: List(root.value) ::: inorderTraversal(root.right)
    }
}

import scala.collection.mutable.{ListBuffer, Stack}

object Solution {
    def inorderTraversal(root: TreeNode): List[Int] = {
        val res = new ListBuffer[Int]()
        val stack = new Stack[TreeNode]()
        var curr = root
        while (curr != null || !stack.isEmpty) {
            while (curr != null) {
                stack.push(curr)
                curr = curr.left
            }
            curr = stack.pop()
            res.append(curr.value)
            curr = curr.right
        }
        res.toList
    }
}

File: Algorithms/Java/0094/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

 // 二叉树中序递归遍历
class Solution {
    private List<Integer> res = new ArrayList<>();

    public List<Integer> inorderTraversal(TreeNode root) {
        inOrder(root);
        return res;
    }

    private void inOrder(TreeNode root) {
        if (root != null) {
            if (root.left != null) {
                inOrder(root.left);
            }
            res.add(root.val);
            if (root.right != null) {
                inOrder(root.right);
            }
        }
    }
}

// 二叉树中序非递归遍历
class Solution {
    public List<Integer> inorderTraversal(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode curr = root;
        while (curr != null || !stack.isEmpty()) {
            while (curr != null) {
                stack.push(curr);
                curr = curr.left;
            }
            curr = stack.pop();
            res.add(curr.val);
            curr = curr.right;
        }
        return res;
    }
}
